service/history/shard/controller.go

```diff
@@ -91,6 +96,7 @@ type (
 
 		sync.RWMutex
 		historyShards map[int]*historyShardsItem
```

Comment:
Since map value is a pointer, it's worth exploring snapshotting on the entire historyShards. Copying here is only copying the addresses of these historyShardsItems so they are low overhead. 

This approach would simplify the code a lot and less error prone. Current approach requires updating snapshot on every historyShards change.

Reply:
Don't all approaches require updating snapshot on every historyShards change? The snapshot is a read-only copy of history shards. If you're referring to store the map in the snapshot, it's not multi-goroutine safe.

Reply:
I was thinking having a single atomic.Pointer field. I've put the suggested patch in the comment if that makes sense.

Reply:
Thanks, but it doesn't seem to be less error prone to me. I prefer to stay with the current approach.